File: app/src/test/java/com/example/financemanager/CashTrackerTests.java
package com.example.financemanager;

import model.Constants;
import controller.CashTracker;
import org.junit.Test;
import org.junit.Assert;

import java.time.LocalDate;

public class CashTrackerTests {
    @Test
    public void comparisonIsCloseEnough() {
        final double x = 100.0;
        final double y = 100.0;
        Assert.assertEquals(x, y, Constants.COMPARISON_DELTA);
    }

    @Test
    public void comparisonIsToFar() {
        final double x = 99.99;
        final double y = 100.0;
        Assert.assertNotEquals(x, y, Constants.COMPARISON_DELTA);
    }

    @Test
    public void testGetMonthsElapsedOverManyMonths() {
        final int addingMonths = 20;

        final LocalDate future = LocalDate.now().minusMonths(20);
        final CashTracker cashTracker = new CashTracker(100.0, future);
        final int monthsElapsed = cashTracker.getMonthsElapsed();
        Assert.assertEquals(addingMonths, monthsElapsed);
    }

    @Test
    public void testGetMonthsElapsedOverNoMonths() {
        final CashTracker cashTracker = new CashTracker(100.0, LocalDate.now());
        final int monthsElapsed = cashTracker.getMonthsElapsed();
        Assert.assertEquals(0, monthsElapsed);
    }

    @Test
    public void trackerPoolIsCorrectOnInit() {
        final CashTracker cashTracker = new CashTracker(100.0, LocalDate.now());
        final double availableFunds = cashTracker.getAvailableFunds();
        Assert.assertEquals(100.0, availableFunds, Constants.COMPARISON_DELTA);
    }

    @Test
    public void canCollectAllRecords() {
        final CashTracker cashTracker = new CashTracker(100.0, LocalDate.now());
        cashTracker.record(20.0);
        cashTracker.record(40.0);
        Assert.assertEquals(60.0, cashTracker.getTotalSpend(), Constants.COMPARISON_DELTA);
    }

    @Test
    public void canAddRecordAndRemovesFunds() {
        final double budget = 100.0;
        final double spend = 20.0;

        final CashTracker cashTracker = new CashTracker(budget, LocalDate.now());
        cashTracker.record(spend);
        Assert.assertEquals(budget - spend, cashTracker.getAvailableFunds(), Constants.COMPARISON_DELTA);
    }
}

File: app/src/test/java/com/example/financemanager/CashTrackerTests.kt
package com.example.financemanager

import model.COMPARISON_DELTA
import model.CashTracker
import org.junit.Test

import org.junit.Assert.*
import java.time.LocalDate

class DoubleComparisonTests {
    @Test
    fun comparisonIsCloseEnough() {
        val x = 100.0;
        val y = 100.0;
        assertEquals(x, y, COMPARISON_DELTA)
    }

    @Test
    fun comparisonIsToFar() {
        val x = 99.99;
        val y = 100.0;
        assertNotEquals(x, y, COMPARISON_DELTA)
    }
}

class TrackerTests {
    @Test
    fun testGetMonthsElapsedOverManyMonths() {
        val addingMonths = 20

        val future = LocalDate.now().minusMonths(20)
        val cashTracker = CashTracker(100.0, future)
        val monthsElapsed = cashTracker.getMonthsElapsed()
        assertEquals(addingMonths, monthsElapsed)
    }

    @Test
    fun testGetMonthsElapsedOverNoMonths() {
        val cashTracker = CashTracker(100.0, LocalDate.now())
        val monthsElapsed = cashTracker.getMonthsElapsed()
        assertEquals(0, monthsElapsed)
    }

    @Test
    fun trackerPoolIsCorrectOnInit() {
        val cashTracker = CashTracker(100.0, LocalDate.now())
        val availableFunds = cashTracker.getAvailableFunds()
        assertEquals(100.0, availableFunds, COMPARISON_DELTA)
    }

    @Test
    fun canCollectAllRecords() {
        val cashTracker = CashTracker(100.0, LocalDate.now())
        cashTracker.record(20.0)
        cashTracker.record(40.0)
        assertEquals(60.0, cashTracker.getTotalSpend(), COMPARISON_DELTA)
    }

    @Test
    fun canAddRecordAndRemovesFunds() {
        val budget = 100.0;
        val spend = 20.0;

        val cashTracker = CashTracker(budget, LocalDate.now())
        cashTracker.record(spend)
        assertEquals(budget - spend, cashTracker.getAvailableFunds(), COMPARISON_DELTA)
    }
}